lib/template-loader.php

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ *
+ * The method returns an empty result for every template so that a 'wp_template' post
+ * is used instead.
+ *
+ * @see gutenberg_find_template
+ *
+ * @param string $template  Path to the template. See locate_template().
+ * @param string $type      Sanitized filename without extension.
+ * @param array  $templates A list of template candidates, in descending order of priority.
+ * @return string Empty string to ensure template file is considered not found.
+ */
+function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
+	global $_wp_current_template_hierarchy;
+
+	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
+		$_wp_current_template_hierarchy = $templates;
+	} else {
+		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
+	}
+
+	return '';
+}
+
+/**
+ * Find the correct 'wp_template' post for the current hierarchy and return the path
+ * to the canvas file that will render it.
+ *
+ * @param string $template_file Original template file. Will be overridden.
+ * @return string Path to the canvas file to include.
+ */
+function gutenberg_find_template( $template_file ) {
+	global $_wp_current_template_post, $_wp_current_template_hierarchy;
+
+	// Bail if no relevant template hierarchy was determined, or if the template file
+	// was overridden another way.
+	if ( ! $_wp_current_template_hierarchy || $template_file ) {
+		return $template_file;
+	}
+
+	$slugs = array_map(
+		'gutenberg_strip_php_suffix',
+		$_wp_current_template_hierarchy
+	);
+
+	// Find most specific 'wp_template' post matching the hierarchy.
+	$template_query = new WP_Query(
+		array(
+			'post_type'      => 'wp_template',
+			'post_status'    => 'publish',
+			'post_name__in'  => $slugs,
+			'orderby'        => 'post_name__in',
+			'posts_per_page' => 1,
+		)
+	);
+
+	if ( $template_query->have_posts() ) {
+		$template_posts            = $template_query->get_posts();
+		$_wp_current_template_post = $template_posts[0];
+	}
+
+	// Add extra hooks for template canvas.
+	add_action( 'wp_head', 'gutenberg_viewport_meta_tag', 0 );
+
+	// This file will be included instead of the theme's template file.
+	return gutenberg_dir_path() . 'lib/template-canvas.php';
+}
+
+/**
+ * Renders the markup for the current template.
+ */
+function gutenberg_render_the_template() {
+	global $_wp_current_template_post;
+
+	if ( ! $_wp_current_template_post || 'wp_template' !== $_wp_current_template_post->post_type ) {
+		echo '<h1>' . esc_html__( 'No matching template found', 'gutenberg' ) . '</h1>';
+		return;
+	}
+
+	$content = $_wp_current_template_post->post_content;
+
+	$content = do_blocks( $content );
+	$content = wptexturize( $content );
+	$content = str_replace( ']]>', ']]&gt;', $content );
+
+	// Wrap block template in .wp-site-blocks to allow for specific descendant styles
+	// (e.g. `.wp-site-blocks > *`).
```

Comment:
Is this necessary? Couldn't the `body` be used instead of `.wp-site-blocks`? Alternatively, the `wp-site-blocks` class could be added to the `body_classes` when a block-based template is being served as opposed to a PHP-based template.

Reply:
`body` will also style `wp_footer` content.

Reply:
Exactly, in a way I agree it'd be nicer to not have this extra div, but the way many themes handle design rules for block content is to use a selector like `.{blocks-wrap-class} > *`, which would indeed style `wp_footer` content that it shouldn't apply to.

I'd say let's be conservative for now and start with this div. We should definitely continue to look for an alternative solution though that satisfies the styling needs.